### PR TITLE
Bugfix/issue 2708 map rect fail

### PR DESCRIPTION
**Pull request by sakrejda**
Proposed fix to map_rect_concurrent.  @wds15 I tried updating the tests to do a range of shard sizes but I'm missing something about the combine step.  Currently I don't feel comfortable saying this is fixed unless we test a range of shard sizes.  It might make more sense to me if you explain the two checks in this test:

```
TEST_F(map_rect, concurrent_ragged_return_size_dd) {
  Eigen::VectorXd res1 = stan::math::map_rect<0, hard_work>(
      shared_params_d, job_params_d, x_r, x_i);

  EXPECT_EQ(res1.size(), 2 * N);

  x_i[1] = std::vector<int>(1, 1);

  Eigen::VectorXd res2 = stan::math::map_rect<1, hard_work>(
      shared_params_d, job_params_d, x_r, x_i);

  EXPECT_EQ(res2.size(), 2 * N + 1);
}
```

## Checklist

- [x] Math issue #1075 

- [x] Copyright holder: Krzysztof Sakrejda

- [X] the basic tests are passing

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [X] the new changes are tested

### CODE REVIEW

**Comment by bob-carpenter** [comment]
I don't see math issue 1708.  Is this addressing #1075?

**Comment by sakrejda** [comment]
Fixed thanks.

**Comment by wds15** [comment]
The idea of these tests is to test if the returned ragged vector has the correct size. The first test is for a case where each job returns 2 elements and the second test is to check if things work if one job (the first one) returns 3 instead of 2 elements.

File: stan/math/prim/mat/functor/map_rect_concurrent.hpp [hunk 1]
```diff
@@ -96,21 +96,25 @@ map_rect_concurrent(
...
+    int job_start = small_job_size;
+    for (int i = 0; i < num_big_threads; ++i) {
+      futures.emplace_back(std::async(std::launch::async, execute_chunk,
+                                      job_start, big_job_size));
+      job_start += big_job_size;
+    }
+    for (int i = 0; i < num_small_threads; ++i) {
+      futures.emplace_back(std::async(std::launch::async, execute_chunk,
+                                      job_start, big_job_size));
+      job_start += small_job_size;
```
Comment (wds15) [inline stan/math/prim/mat/functor/map_rect_concurrent.hpp]: I like how you code it, but I think that you intended to schedule in the first for loop all the small jobs and in the second for loop all the big jobs. Also note that right now you schedule in both for loops `big_job_size` - but that should be different in the two loops (first the small, then the big ones).

**Comment by sakrejda** [comment]
@wds15 can you answer my question up in the main PR description about the 'sizes' test? Why does changing the integer parameters (or running the command a second time with another id) increment the return size by 1? I'll look at it again now but if you have a quick answer to that question it would save me time with adding tests.

**Comment by sakrejda** [comment]
re: joint testing with MPI, I agree that's a good idea, would also be good to separate the job counting out into shared code between the two and use a simple thread pool here (it's not bad with c++1y).. I'm going to leave those for later as this is a bugfix PR.  Definitely want more tests, we should've tested for varying shard/thread numbers to begin with. :/

**Comment by sakrejda** [comment]
re: functor ok, no idea how I missed that, thanks.  I think I can do the tests now.

**Comment by sakrejda** [comment]
@syclik how is the number of threads set for testing? I think we should expand the tests by running the current tests with a range of STAN_NUM_THREADS values but that doesn't seem to be in the testing framework itself.  Is that set via Jenkins or something like that?

**Comment by wds15** [comment]
Have a look into the get num threads test. There the variable is set to different values.

**Comment by wds15** [comment]
@sakrejda I am very close to have a PR ready which would allow us to use the parallel STL from Intel as backend in addition to the async facility. This will give us nice speedups - in particular when jobs have unequal cpu needs as the Parallel STL from Intel can deal with this.

This PR would, of course, contain a fix for this issue as well. The downside that this entails a few more lines to review, but not very hard.

The Parallel STL from Intel is rather easy to install via their binary distribution on Mac...and there is even hope for Windows, I think.

**Comment by wds15** [comment]
uh... sorry to hear, but while not blessed with fever, I also do enjoy right now the season...

**Comment by sakrejda** [comment]
@wds15 I think we should keep this branch as a bugfix rather than letting it creep.  I'm happy to review the parallel STL updates on a separate feature PR, just trying to get this in asap esp. since I've been out.

**Comment by sakrejda** [comment]
There's still some segfault in here, with 5 shards and 4 threads.

**Comment by wds15** [comment]
This PR is not quote right yet, I think.

We can work on this one, but the other parallel STL one has all fixes and adequate tests.

If we continue this one, we should at least take over some pieces (like the test) from the parallel STL. However, the parallel STL PR is not too massive ... all I am doing there is to refactor `map_rect_concurrent` to use `for_each` from C++17 (and providing a C++11 implementation of that).

**Comment by syclik** [comment]
@wds15, please work on a fix first before any new improvements. Given that
this feature has lead to a major bug, this is high priority. And new
features will have to be evaluated carefully.

**Comment by sakrejda** [comment]
@wds15 @syclik I'm closing this.  @wds15 Can you just take the relevant parts of the STL PR and do a new bugfix PR?

**Comment by syclik** [comment]
@sakrejda, thanks for letting me know. Feel better.

@wds15, agreed. If you've fixed it on a branch that has a new feature, please separate out the fix so we can patch the issue immediately.

**Pull request closed without merging**